Support AWS IMDSv2

arossert opened the pull request:
Added support for IMDSv2
https://github.com/dgzlopes/cloud-detect/issues/27

kshivakumar reviewed (commented):
@arossert changes look fine. Since IMDSv2 is supposed to be more secure than v1, it's good to implement it now.
Have you tested it in both Linux and Windows EC2 instances?
I couldn't test it myself as I am having some issues with my AWS account, haven't used it in long time.

kshivakumar commented:
@arossert Can you call the two APIs concurrently using `asyncio.gather`.
As per the [documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html), each instance is user-configured to use only one of the APIs. It seems in your case, your Windows EC2 instances are configured to use v2.

arossert commented:
@kshivakumar yes I tested it on both Windows and Linux, what I don't know to tell is if V2 can be disabled and `api/token` call can fail.

If this is the case we can try to do what you suggested and do both, one to try with V1 (without token) and the other with.

Let me know if you think it will work better

kshivakumar commented:
> I tested it on both Windows and Linux...

Great!
@arossert Yeah, please go ahead with concurrently calling both the APIs.

arossert commented:
@kshivakumar I ended up keeping the old `_get_metadata` flow and added `_get_metadata_v2` that generates a token and calls the `_get_metadata` with the needed header.

I removed the try...catch block and just ignored the exceptions with the `return_exceptions` flag.
I will run some more tests on this new changes and report back

Let me know if this looks good to you.

kshivakumar commented on cloud_detect/providers/aws_provider.py:
```diff
@@ -28,22 +32,34 @@ async def identify(self):
 
+    async def _get_metadata_v2(self):
+        async with aiohttp.ClientSession() as session:
+            async with session.put(self.metadata_token_url, headers={'X-aws-ec2-metadata-token-ttl-seconds': '21600'}) as response:
```
A ttl of one minute is sufficient for the library to make the second api call.

arossert commented on cloud_detect/providers/aws_provider.py:
```diff
@@ -28,22 +32,34 @@ async def identify(self):
+        results = await asyncio.gather(
+            self._get_metadata(),
+            self._get_metadata_v2(),
+            return_exceptions=True
```
I added the `return_exceptions` because I think that it will make the code look cleaner, but I will add the try block instead if you prefer.
The `raise_for_status` is needed since the API call can return with 401 status and an empty body, I assume that `response.json` will also fail in this case 😃

arossert replied:
Another option is to replace the try block with `contextlib.suppress(BaseException)`
BTW, any reason you are using `BaseException` and not `Exception`?

kshivakumar replied:
@arossert 
> I added the return_exceptions because I think that it will make the code look cleaner...

I guessed it, I did the same some time back 😄 
`return_exceptions=True` gives an impression of processing the exception object later, which we are not doing here. 

> The raise_for_status is needed since the API call can return with 401 status and an empty body, I assume that response.json will also fail in this case ...

Yeah, checking the json response automatically covers all the scenarios.

> Another option is to replace the try block with contextlib.suppress(BaseException)

I was not aware of that, looks cool. You can implement it.

> BTW, any reason you are using BaseException and not Exception?

That's how it was from the [beginning](https://github.com/dgzlopes/cloud-detect/commit/1611e77ee92f17d92f0b07e16636e32a83763343), I didn't give much thought either when I introduced `asyncio`.
The only advantage I see is that it stops nicely without any traceback and returns `False` when cancelled by user or another process. The downside is false positive when interrupted.

arossert replied:
Thanks for the information, it seems a bit weird to me since `BaseExceptoin` is rarely used.
Anyway, I have replaced the `try` with `contextlib`, if you are happy with this change I would appreciate it if you would merge and release a new version 🙏

kshivakumar commented:
@arossert Please fix the "DCO" check.

arossert commented:
> @arossert Please fix the "DCO" check.

Fixed
